MessageStateControllerTest is unstable

Stack trace:

```
[ERROR] Tests run: 22, Failures: 1, Errors: 0, Skipped: 1, Time elapsed: 3.726 s <<< FAILURE! - in io.zeebe.broker.subscription.message.state.MessageStateControllerTest
[ERROR] shouldFindMessageBeforeTime(io.zeebe.broker.subscription.message.state.MessageStateControllerTest)  Time elapsed: 0.126 s  <<< FAILURE!
org.junit.ComparisonFailure: expected:<[1]> but was:<[0]>
        at sun.reflect.NativeConstructorAccessorImpl.newInstance0(Native Method)
        at sun.reflect.NativeConstructorAccessorImpl.newInstance(NativeConstructorAccessorImpl.java:62)
        at sun.reflect.DelegatingConstructorAccessorImpl.newInstance(DelegatingConstructorAccessorImpl.java:45)
        at io.zeebe.broker.subscription.message.state.MessageStateControllerTest.shouldFindMessageBeforeTime(MessageStateControllerTest.java:303)
        at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
        at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
        at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
        at java.lang.reflect.Method.invoke(Method.java:498)
        at org.junit.runners.model.FrameworkMethod$1.runReflectiveCall(FrameworkMethod.java:50)
        at org.junit.internal.runners.model.ReflectiveCallable.run(ReflectiveCallable.java:12)
        at org.junit.runners.model.FrameworkMethod.invokeExplosively(FrameworkMethod.java:47)
        at org.junit.internal.runners.statements.InvokeMethod.evaluate(InvokeMethod.java:17)
        at org.junit.internal.runners.statements.RunBefores.evaluate(RunBefores.java:26)
        at org.junit.internal.runners.statements.RunAfters.evaluate(RunAfters.java:27)
        at org.junit.rules.ExternalResource$1.evaluate(ExternalResource.java:48)
        at org.junit.rules.RunRules.evaluate(RunRules.java:20)
        at org.junit.runners.ParentRunner.runLeaf(ParentRunner.java:325)
        at org.junit.runners.BlockJUnit4ClassRunner.runChild(BlockJUnit4ClassRunner.java:78)
        at org.junit.runners.BlockJUnit4ClassRunner.runChild(BlockJUnit4ClassRunner.java:57)
        at org.junit.runners.ParentRunner$3.run(ParentRunner.java:290)
        at org.junit.runners.ParentRunner$1.schedule(ParentRunner.java:71)
        at org.junit.runners.ParentRunner.runChildren(ParentRunner.java:288)
        at org.junit.runners.ParentRunner.access$000(ParentRunner.java:58)
        at org.junit.runners.ParentRunner$2.evaluate(ParentRunner.java:268)
        at org.junit.runners.ParentRunner.run(ParentRunner.java:363)
        at org.junit.runners.Suite.runChild(Suite.java:128)
        at org.junit.runners.Suite.runChild(Suite.java:27)
        at org.junit.runners.ParentRunner$3.run(ParentRunner.java:290)
        at org.junit.runners.ParentRunner$1.schedule(ParentRunner.java:71)
        at org.junit.runners.ParentRunner.runChildren(ParentRunner.java:288)
        at org.junit.runners.ParentRunner.access$000(ParentRunner.java:58)
        at org.junit.runners.ParentRunner$2.evaluate(ParentRunner.java:268)
        at org.junit.runners.ParentRunner.run(ParentRunner.java:363)
        at org.apache.maven.surefire.junitcore.JUnitCore.run(JUnitCore.java:55)
        at org.apache.maven.surefire.junitcore.JUnitCoreWrapper.createRequestAndRun(JUnitCoreWrapper.java:137)
        at org.apache.maven.surefire.junitcore.JUnitCoreWrapper.executeEager(JUnitCoreWrapper.java:107)
        at org.apache.maven.surefire.junitcore.JUnitCoreWrapper.execute(JUnitCoreWrapper.java:83)
        at org.apache.maven.surefire.junitcore.JUnitCoreWrapper.execute(JUnitCoreWrapper.java:75)
        at org.apache.maven.surefire.junitcore.JUnitCoreProvider.invoke(JUnitCoreProvider.java:158)
        at org.apache.maven.surefire.booter.ForkedBooter.invokeProviderInSameClassLoader(ForkedBooter.java:383)
        at org.apache.maven.surefire.booter.ForkedBooter.runSuitesInProcess(ForkedBooter.java:344)
        at org.apache.maven.surefire.booter.ForkedBooter.execute(ForkedBooter.java:125)
        at org.apache.maven.surefire.booter.ForkedBooter.main(ForkedBooter.java:417)

```

--- a/broker-core/src/main/java/io/zeebe/broker/subscription/message/processor/PublishMessageProcessor.java
+++ b/broker-core/src/main/java/io/zeebe/broker/subscription/message/processor/PublishMessageProcessor.java
@@ -32,6 +32,7 @@ import io.zeebe.broker.subscription.message.state.MessageSubscription;
 import io.zeebe.protocol.clientapi.RejectionType;
 import io.zeebe.protocol.impl.record.value.message.MessageRecord;
 import io.zeebe.protocol.intent.MessageIntent;
+import io.zeebe.util.sched.clock.ActorClock;
 import java.util.List;
 import java.util.function.Consumer;
 
@@ -96,7 +97,8 @@ public class PublishMessageProcessor implements TypedRecordProcessor<MessageReco
                 messageRecord.getCorrelationKey(),
                 messageRecord.getPayload(),
                 messageRecord.getMessageId(),
-                messageRecord.getTimeToLive());
+                messageRecord.getTimeToLive(),
+                messageRecord.getTimeToLive() + ActorClock.currentTimeMillis());
         messageStateController.put(message);
 
       } else {

--- a/broker-core/src/main/java/io/zeebe/broker/subscription/message/state/Message.java
+++ b/broker-core/src/main/java/io/zeebe/broker/subscription/message/state/Message.java
@@ -23,7 +23,6 @@ import static io.zeebe.util.buffer.BufferUtil.writeIntoBuffer;
 
 import io.zeebe.util.buffer.BufferReader;
 import io.zeebe.util.buffer.BufferWriter;
-import io.zeebe.util.sched.clock.ActorClock;
 import org.agrona.DirectBuffer;
 import org.agrona.MutableDirectBuffer;
 import org.agrona.concurrent.UnsafeBuffer;
@@ -46,7 +45,8 @@ public final class Message implements BufferWriter, BufferReader {
       DirectBuffer correlationKey,
       DirectBuffer payload,
       DirectBuffer id,
-      long timeToLive) {
+      long timeToLive,
+      long deadline) {
     this.name.wrap(name);
     this.correlationKey.wrap(correlationKey);
     this.payload.wrap(payload);
@@ -54,7 +54,7 @@ public final class Message implements BufferWriter, BufferReader {
 
     this.key = key;
     this.timeToLive = timeToLive;
-    this.deadline = ActorClock.currentTimeMillis() + timeToLive;
+    this.deadline = deadline;
   }
 
   public DirectBuffer getName() {

--- a/broker-core/src/test/java/io/zeebe/broker/subscription/message/state/MessageStateControllerTest.java
+++ b/broker-core/src/test/java/io/zeebe/broker/subscription/message/state/MessageStateControllerTest.java
@@ -277,20 +277,15 @@ public class MessageStateControllerTest {
   @Test
   public void shouldNotFindMessageBeforeTime() {
     // given
-    final long now = ActorClock.currentTimeMillis();
-
     final Message message = createMessage(1L, "name", "correlationKey", "{}", "nr1", 1234);
     final Message message2 = createMessage(2L, "name", "correlationKey", "{}", "nr2", 4567);
 
     stateController.put(message);
     stateController.put(message2);
 
-    // when
-    final long deadline = now + 1_000L;
-
     // then
     final List<Message> readMessage = new ArrayList<>();
-    stateController.findMessagesWithDeadlineBefore(deadline, readMessage::add);
+    stateController.findMessagesWithDeadlineBefore(1_000, readMessage::add);
 
     assertThat(readMessage).isEmpty();
   }
@@ -320,20 +315,15 @@ public class MessageStateControllerTest {
   @Test
   public void shouldFindMessageBeforeTime() {
     // given
-    final long now = ActorClock.currentTimeMillis();
-
     final Message message = createMessage(1L, "name", "correlationKey", "{}", "nr1", 1234);
     final Message message2 = createMessage(2L, "otherName", "correlationKey", "{}", "nr2", 2000);
 
     stateController.put(message);
     stateController.put(message2);
 
-    // when
-    final long deadline = now + 1999L;
-
     // then
     final List<Message> readMessage = new ArrayList<>();
-    stateController.findMessagesWithDeadlineBefore(deadline, readMessage::add);
+    stateController.findMessagesWithDeadlineBefore(1_999, readMessage::add);
 
     assertThat(readMessage.size()).isEqualTo(1);
     assertThat(readMessage.get(0).getKey()).isEqualTo(1L);
@@ -626,13 +616,20 @@ public class MessageStateControllerTest {
         wrapString(correlationKey),
         wrapString(""),
         new UnsafeBuffer(0, 0),
+        10_000,
         0L);
   }
 
   private Message createMessage(
       long key, String name, String correlationKey, String payload, String id) {
     return new Message(
-        key, wrapString(name), wrapString(correlationKey), wrapString(payload), wrapString(id), 0L);
+        key,
+        wrapString(name),
+        wrapString(correlationKey),
+        wrapString(payload),
+        wrapString(id),
+        10_000,
+        0L);
   }
 
   private Message createMessage(
@@ -643,6 +640,7 @@ public class MessageStateControllerTest {
         wrapString(correlationKey),
         wrapString(payload),
         wrapString(id),
+        10_000L,
         deadline);
   }
 }

